pkg/pattern/tokenization/tokenization.go

```diff
@@ -0,0 +1,1054 @@
+		return true
+
+	// The number isn't followed by a boundary char (which include things like
+	// '.', ' ', '/', etc.), so it's either not a real number or date, or it's
+	// some sort of a unit like a duration (e.g. 3h5m2s), size, (e.g. 12KB,
+	// 3MiB), etc.
+	case !boundryChars[b1]:
+		return r.handleHexOrUnit(hasMinusPrefix, n1, l1, b1)
+
+	// We have a decimal point, so this can either be a plain number, a unit
+	// that can have a float value, or an IP address.
+	case b1 == '.':
+		return r.handleNumberWithDecimal(hasMinusPrefix, n1, l1)
+
+	// This might be a timestamp that looks like '2024-04-01...' or
+	// `2024/03/27...`; timestamps in the far future are not supported :)
+	case n1 <= maxYear && l1 <= 4 && (b1 == '-' || b1 == '/'):
+		return r.handleSaneTimestamp(hasMinusPrefix, n1, b1)
+
+	// Weird RFC822 dates like "02 Jan 06 15:04 MST"
+	case n1 <= 31 && l1 <= 2 && b1 == ' ':
+		if r.advanceMonthName() && r.advanceChar(' ') && r.advanceYear() && r.advanceChar(' ') && r.advanceTime(true) && r.advanceStringOrNumericTimeZone(false) {
+			r.commit()
+			r.emit(hasMinusPrefix, placeholderTimestamp)
+			return true
+		}
+		// if not, go to default handler after switch statement
+
+	// It could be a UUID that starts with 8 digits
+	case l1 == 8 && b1 == '-':
+		if r.advanceUUIDAfterFirstDash('a', 'f') || (r.resetHeadExpr() && r.advanceChar('-') && r.advanceUUIDAfterFirstDash('A', 'F')) {
+			r.emit(hasMinusPrefix, placeholderUUID)
+			return true
+		}
+		// if not, go to default handler after switch statement
+	}
+
+	// Number with an unknown boundary - emit the number and leave the boundary
+	// for the following passes.
+	r.resetHead()
+	r.emitNumberOrCopyText(hasMinusPrefix)
+	return true
+}
+
+var longDayNames = [...][]byte{
+	[]byte("Sunday"),
+	[]byte("Monday"),
+	[]byte("Tuesday"),
+	[]byte("Wednesday"),
+	[]byte("Thursday"),
+	[]byte("Friday"),
+	[]byte("Saturday"),
+}
+
+func (r *replacer) handleWeirdTimestamp() (endsWithBoundary bool) {
+	r.resetHead()
+
+	if r.advanceDayOfTheWeek() {
+		r.commit() // we will always consume this
+
+		// RFC1123 and RFC1123Z, e.g.:
+		// - "Mon, 02 Jan 2006 15:04:05 MST"
+		// - "Mon, 02 Jan 2006 15:04:05 -0700"
+		if r.advanceChar(',') && r.advanceChar(' ') && r.advanceUintUpTo(31, 2) && r.advanceChar(' ') && r.advanceMonthName() && r.advanceChar(' ') && r.advanceYear() && r.advanceChar(' ') && r.advanceTime(false) && r.advanceStringOrNumericTimeZone(false) {
+			r.emit(false, placeholderTimestamp)
+			return true
+		}
+		r.resetHead()
+
+		// ANSIC, UnixDatem, RubyDate e.g
+		// - "Mon Jan  2 15:04:05 2006"
+		// - "Mon Jan  2 15:04:05 MST 2006"
+		// - "Mon Jan 02 15:04:05 -0700 2006"
+		if r.advanceChar(' ') && r.advanceMonthName() && r.advanceChar(' ') && (r.advanceChar(' ') || true) && r.advanceUintUpTo(31, 2) && r.advanceChar(' ') && r.advanceTime(false) && r.advanceStringOrNumericTimeZone(true) && r.advanceChar(' ') && r.advanceYear() {
```

Comment:
`&& (r.advanceChar(' ') || true)` this is clever 😻

Reply:
One of the potential minor performance improvements I want to benchmark at some future point is actually getting rid of `advanceOneOf`. We only use it in a couple of places, here: https://github.com/grafana/loki/blob/c62c6cfb746392a31e9dad52cd04cca4bf7cd152/pkg/pattern/tokenization/replacer.go#L191
and here: https://github.com/grafana/loki/blob/c62c6cfb746392a31e9dad52cd04cca4bf7cd152/pkg/pattern/tokenization/replacer.go#L535

And both of these can be easily rewritten like this:
```go
(r.advanceChar('+') || r.advanceChar('-'))
(r.advanceChar('T') || r.advanceChar(' '))
```

I'll actually comment on a few other potential minor performance improvements I've thought of but haven't had time to benchmark, just so I don't forget them in the future :sweat_smile: